src/Compilers/Core/Portable/DocumentationCommentId.cs

```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
             return name;
         }
 
-        private class DeclarationGenerator : SymbolVisitor
+        /// <summary>
+        /// The <see langword="bool"/> type parameter indicates if we succeeded at writing out the declaration ID or not. 
+        /// </summary>
+        private sealed class PrefixAndDeclarationGenerator(StringBuilder builder) : SymbolVisitor<bool>
```

Comment:
>PrefixAndDeclarationGenerator [](http://example.com/codeflow?start=29&length=29)

I find changes made to the visitors both fragile an unnecessary. Therefore here is what I suggest to do:
1. Revert all changes in this file, including all renames and unrelated optimizations.  
2. Rather than throwing ```InvalidOperationException```, switch to throwing a private exception type in the visitors
3. In the two/three places that use the visitors, catch the private exception and assume the result is null.
4. If you still would like to rename types and optimize StringBuilder operations, submit those as separate PRs.

Reply:
I would prefer to not revert the renames.  They were very helpful for actually determining what the type was doing and seeing where things were odd.  Keepign clear names (including the distinction on if the DocID prefix is included or not very much helped with clarity and diagnosability here.

2. I would prefer to not use exceptions in normal non-exceptional codepaths.  They are expensive under a debugger, and often end up interfering with teh debugging experience (breaking during normal operations).  We get reports from VS and other teams about pain points here when exceptions are thrown in codepaths that done need it.

3. There seems to be only one real codepath that flows null upwards.  And it was clearly identified by either being DefaultVisit, or a recursive call to Visit.  This was easy to audit and write code for.  So i'd prefer to just use flowing that boolean upwards versus needing to use an exception with all the machinery that entails.

4. I've pulled out the stringbuilder change.  I can do that in anotehr PR.  HOwever, the renames and other structural changes were very relevant in the course of auditing this type.  IMO, had those changes been there originally, i think it would have been less likely that these issues would have arisen. As such, i do not view them as stylistic, but as inherent to good design and impl for this space.

Reply:
Sorry, but my position didn't change. We can discuss offline in more details.

Reply:
Other auditing parts have been pulled out to a separate pr.